cloud/wfm/src/executor/ExecutorImpl.ts

```diff
@@ -0,0 +1,57 @@
+import * as Promise from 'bluebird';
+import ProcessInstance from '../process-instance/ProcessInstance';
+import Process from '../process/Process';
+import Task, {TaskEventData, TaskStatus} from '../task/Task';
+import Executor from './Executor';
+import ExecutorRepository from './ExecutorRepository';
+
+/**
+ * Default implementation for {@link Executor}
+ */
+export default class ExecutorImpl implements Executor {
+  public instance: ProcessInstance;
+  /**
+   * @param process The {@link Process} this {@link Executor} is to run
+   * @param repository {@link ExecutorRepository} instance for data access
+   */
+  constructor(public process: Process, public repository: ExecutorRepository) {
+    this.instance = this.process.createInstance();
+  }
+  public start() {
+    this.runCurrentTask();
+  }
+  protected onTaskDone(e: TaskEventData<Task>) {
+    if (e.currentStatus === TaskStatus.DONE) {
+      Promise.resolve(this.instance.nextTask())
+        .then(() => this.saveInstance())
+        .then(() => this.runCurrentTask());
+    }
+  }
+
+  /**
+   * Orders {@link repository} to update {@link instance}
+   */
+  protected saveInstance() {
+    return this.repository.saveInstance(this.instance);
+  }
+
+  /**
+   * Runs the present {@link Task} of {@link instance} and sets up execution of
+   * the next one as it is run
+   */
+  protected runCurrentTask() {
+    this.instance.getCurrentTask()
+    .then(task => {
+      // currentTask is undefined after all are done
+      if (!task) {
+        return;
+      }
+      return task
+        // TODO: this assumes the Task object will be kept in memory forever
+        // but human-executed tasks can take very long to complete
+        // we might have to change this to an out-of-process message bus
```

Comment:
In memory? Message bus? This makes me worried about what we building here.
This code should execute on the client.
Everything will be sync based. Storage should be represented by repositories. 
Tasks will be stored on the client.
Worth to talk about this to avoid confusion.